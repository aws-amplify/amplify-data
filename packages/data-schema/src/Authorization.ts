import type { UnionToIntersection } from '@aws-amplify/data-schema-types';

const __data = Symbol('data');

/**
 * All possible providers.
 *
 * This list should not be used if you need to restrict available providers
 * according to an auth strategcy. E.g., `public` auth can only be facilitated
 * by `apiKey` and `iam` providers.
 */
export const Providers = [
  'apiKey',
  'iam',
  'userPools',
  'oidc',
  'function',
] as const;
export type Provider = (typeof Providers)[number];

/**
 * The subset of auth providers that can facilitate `public` auth.
 */
export const PublicProviders = ['apiKey', 'iam'] as const;
export type PublicProvider = (typeof PublicProviders)[number];

/**
 * The subset of auth providers that can facilitate `private` auth.
 */
export const PrivateProviders = ['userPools', 'oidc', 'iam'] as const;
export type PrivateProvider = (typeof PrivateProviders)[number];

/**
 * The subset of auth providers that can facilitate `owner` auth.
 */
export const OwnerProviders = ['userPools', 'oidc'] as const;
export type OwnerProviders = (typeof OwnerProviders)[number];

/**
 * The subset of auth providers that can facilitate `group` auth.
 */
export const GroupProviders = ['userPools', 'oidc'] as const;
export type GroupProvider = (typeof GroupProviders)[number];

/**
 * The subset of auth providers that can facilitate `custom` auth.
 */
export const CustomProviders = ['function'] as const;
export type CustomProvider = (typeof CustomProviders)[number];

export const Strategies = [
  'public',
  'private',
  'owner',
  'groups',
  'custom',
] as const;
export type Strategy = (typeof Strategies)[number];

/**
 * The operations that can be performed against an API.
 */
export const Operations = [
  'create',
  'update',
  'delete',
  'read',
  'get',
  'list',
  'sync',
  'listen',
  'search',
] as const;
export type Operation = (typeof Operations)[number];

export type Authorization<
  AuthStrategy extends Strategy,
  AuthField extends string | undefined,
  AuthFieldPlurality extends boolean,
> = {
  [__data]: {
    strategy?: AuthStrategy;
    provider?: Provider;
    operations?: Operation[];
    groupOrOwnerField?: AuthField;
    groups?: string[];
    multiOwner: AuthFieldPlurality;
    identityClaim?: string;
    groupClaim?: string;
  };
};

export type OwnerField = object;

type BuilderMethods<T extends object> = {
  [K in keyof T as T[K] extends (...args: any) => any ? K : never]: T[K];
};

/**
 * Creates a shallow copy of an object with an individual field pruned away.
 *
 * @param original The original object to prune.
 * @param without The field to prune.
 * @returns The pruned object.
 */
function omit<T extends object, O extends string>(
  original: T,
  without: O,
): Omit<T, O> {
  const pruned = { ...original };
  delete (pruned as any)[without];
  return pruned;
}

function to<SELF extends Authorization<any, any, any>>(
  this: SELF,
  operations: Operation[],
) {
  (this as any)[__data].operations = operations;
  return omit(this, 'to');
}

function inField<
  SELF extends Authorization<any, any, any>,
  Field extends string,
>(this: SELF, field: Field) {
  this[__data].groupOrOwnerField = field;
  const built = omit(this, 'inField');

  return built as unknown as BuilderMethods<typeof built> &
    Authorization<
      SELF[typeof __data]['strategy'],
      Field,
      SELF[typeof __data]['multiOwner']
    >;
}

/**
 * Specifies a property of the identity JWT to use in place of `sub::username`
 * as the value to match against the owner field for authorization.
 *
 * @param this Authorization object to operate against.
 * @param property A property of identity JWT.
 * @returns A copy of the Authorization object with the claim attached.
 */
function identityClaim<SELF extends Authorization<any, any, any>>(
  this: SELF,
  property: string,
) {
  this[__data].identityClaim = property;
  return omit(this, 'identityClaim');
}

function withClaimIn<SELF extends Authorization<any, any, any>>(
  this: SELF,
  property: string,
) {
  this[__data].groupClaim = property;
  return omit(this, 'withClaimIn');
}

function validateProvider(
  needle: Provider | undefined,
  haystack: readonly Provider[],
) {
  if (needle && !haystack.includes(needle)) {
    throw new Error(`Invalid provider (${needle}) given!`);
  }
}

function authData<
  Strat extends Strategy = 'public',
  Field extends string | undefined = undefined,
  isMulti extends boolean = false,
  Builders extends object = object,
>(
  defaults: Partial<Authorization<Strat, Field, isMulti>[typeof __data]>,
  builderMethods: Builders,
): Authorization<Strat, Field, isMulti> & Builders {
  return {
    [__data]: {
      strategy: 'public',
      provider: undefined,
      operations: undefined,
      groupOrOwnerField: undefined,
      multiOwner: false,
      identityClaim: undefined,
      groups: undefined,
      ...defaults,
    } as any,
    ...builderMethods,
  };
}

/**
 * Defines an authorization rule for your data models and fields. First choose an authorization strategy (`public`,
 * `private`, `owner`, `group`, or `custom`), then choose an auth provider (`apiKey`, `iam`, `userPools`, `oidc`, or `function`)
 * and optionally use `.to(...)` to specify the operations that can be performed against your data models and fields.
 */
export const allow = {
  /**
   * Authorize unauthenticated users. By default, `.public()` uses an API key based authorization. You can additionally
   * use `.public('iam')` to use IAM based authorization for unauthenticated users.
   * @param provider the authentication provider - supports "apiKey" or "iam" as valid providers
   * @returns an authorization rule for unauthenticated users
   */
  public(provider?: PublicProvider) {
    validateProvider(provider, PublicProviders);
    return authData(
      {
        strategy: 'public',
        provider,
      },
      {
        to,
      },
    );
  },

  /**
   * Authorize authenticated users. By default, `.private()` uses an Amazon Cognito user pool based authorization. You can additionally
   * use `.private("iam")` or `.private("oidc")` to use IAM or OIDC based authorization for authenticated users.
   * @param provider the authentication provider - supports "userPools", "iam", or "oidc"
   * @returns an authorization rule for authenticated users
   */
  private(provider?: PrivateProvider) {
    validateProvider(provider, PrivateProviders);
    return authData(
      {
        strategy: 'private',
        provider,
      },
      {
        to,
      },
    );
  },

  /**
   * Authorize access on a per-user (owner) basis. By setting owner-based authorization, a new `owner: a.string()`
   * field will be added to the model to store which user "owns" the item. Upon item creation, the "owner field" is
   * auto-populated with the authenticated user's information. You can which field should be used as the owner field
   * by chaining the * `.inField(...)` method.
   *
   * By default, `.owner()` uses an Amazon Cognito user pool based authorization. You can additionally
   * use `.owner("oidc")` to use OIDC based authentication to designate the owner.
   *
   * To change the specific claim that should be used as the user identifier within the owner field, chain the
   * `.identityClaim(...)` method.
   *
   * @param provider the authentication provider - supports "userPools", "iam", or "oidc"
   * @returns an authorization rule for authenticated users
   */
  owner(provider?: OwnerProviders) {
    validateProvider(provider, OwnerProviders);
    return authData(
      {
        strategy: 'owner',
        provider,
        groupOrOwnerField: 'owner',
      },
      {
        to,
        inField,
        identityClaim,
      },
    );
  },

  /**
   * Authorize access for multi-user / multi-owner access. By setting multi-owner-based authorization, a new `owners: a.string().array()`
   * field will be added to the model to store which users "own" the item. Upon item creation, the "owners field" is
   * auto-populated with the authenticated user's information. To grant other users access to the item, append their user identifier into the `owners` array.
   * You can which field should be used as the owners field by chaining the * `.inField(...)` method.
   *
   * By default, `.multipleOwners()` uses an Amazon Cognito user pool based authorization. You can additionally
   * use `.multipleOwners("oidc")` to use OIDC based authentication to designate the owner.
   *
   * To change the specific claim that should be used as the user identifier within the owners field, chain the
   * `.identityClaim(...)` method.
   *
   * @param provider the authentication provider - supports "userPools", "iam", or "oidc"
   * @returns an authorization rule for authenticated users
   */
  multipleOwners(provider?: OwnerProviders) {
    validateProvider(provider, OwnerProviders);
    return authData(
      {
        strategy: 'owner',
        multiOwner: true,
        provider,
        groupOrOwnerField: 'owner',
      },
      {
        to,
        inField,
        identityClaim,
      },
    );
  },

  /**
   * Authorize a specific user group. Provide the name of the specific user group to have access.
   *
   * By default, `.specificGroup()` uses an Amazon Cognito user pool based authorization. You can additionally
   * use `.specificGroup("group-name", "oidc")` to use OIDC based authentication to designate the user group.
   *
   * To change the specific claim that should be used as the user group identifier, chain the
   * `.withClaimIn(...)` method.
   * @param group the name of the group to authorize
   * @param provider the authentication provider - supports "userPools" or "oidc"
   * @returns an authorization rule to grant access by a specific group
   */
  specificGroup(group: string, provider?: GroupProvider) {
    return authData(
      {
        strategy: 'groups',
        provider,
        groups: [group],
      },
      {
        to,
        withClaimIn,
      },
    );
  },

  /**
   * Authorize multiple specific user groups. Provide the names of the specific user groups to have access.
   *
   * By default, `.specificGroups()` uses an Amazon Cognito user pool based authorization. You can additionally
   * use `.specificGroups(["group-a", "group-b"], "oidc")` to use OIDC based authentication to designate the user group.
   *
   * To change the specific claim that should be used as the user group identifier, chain the
   * `.withClaimIn(...)` method.
   * @param groups the names of the group to authorize defined as an array
   * @param provider the authentication provider - supports "userPools" or "oidc"
   * @returns an authorization rule to grant access by a specific group
   */
  specificGroups(groups: string[], provider?: GroupProvider) {
    return authData(
      {
        strategy: 'groups',
        provider,
        groups,
      },
      {
        to,
        withClaimIn,
      },
    );
  },

  /**
   * Authorize if a user is part of a group defined in a data model field.
   *
   * By default, `.groupDefinedIn()` uses an Amazon Cognito user pool based authorization. You can additionally
   * use `.groupDefinedIn("field-name", "oidc")` to use OIDC based authentication to designate the user group.
   *
   * To change the specific claim that should be used as the user group identifier within the groups field, chain the
   * `.withClaimIn(...)` method.
   * @param groupsField the field that should store the authorized user group information
   * @param provider the authentication provider - supports "userPools" or "oidc"
   * @returns an authorization rule to grant access by a specific group
   */
  groupDefinedIn<T extends string>(groupsField: T, provider?: GroupProvider) {
    return authData(
      {
        strategy: 'groups',
        provider,
        groupOrOwnerField: groupsField,
      },
      {
        to,
      },
    );
  },

  /**
   * Authorize if a user is part of a one of the groups defined in a data model field.
   *
   * By default, `.groupsDefinedIn()` uses an Amazon Cognito user pool based authorization. You can additionally
   * use `.groupsDefinedIn("field-name", "oidc")` to use OIDC based authentication to designate the user group.
   *
   * To change the specific claim that should be used as the user group identifier within the groups field, chain the
   * `.withClaimIn(...)` method.
   * @param groupsField the field that should store the list of authorized user groups
   * @param provider the authentication provider - supports "userPools" or "oidc"
   * @returns an authorization rule to grant access by a specific group
   */
  groupsDefinedIn<T extends string>(groupsField: T, provider?: GroupProvider) {
    return authData(
      {
        strategy: 'groups',
        provider,
        groupOrOwnerField: groupsField,
        multiOwner: true,
      },
      {
        to,
      },
    );
  },

  custom(provider?: CustomProvider) {
    return authData(
      {
        strategy: 'custom',
        provider,
      },
      {
        to,
      },
    );
  },
} as const;

/**
 * Turns the type from a list of `Authorization` rules like this:
 *
 * ```typescript
 * [
 *  allow.public(),
 *  allow.owner().inField('otherfield'),
 *  allow.multipleOwners().inField('editors')
 * ]
 * ```
 *
 * Into a union of the possible `fieldname: type` auth objects like this:
 *
 * ```typescript
 * {
 *  owner?: string | undefined;
 * } | {
 *  otherfield?: string | undefined;
 * } | {
 *  editors?: string[] | undefined;
 * }
 * ```
 */
export type ImpliedAuthField<T extends Authorization<any, any, any>> =
  T extends Authorization<infer _Strat, infer Field, infer isMulti>
    ? Field extends undefined
      ? never
      : Field extends string
      ? isMulti extends true
        ? { [K in Field]?: string[] }
        : { [K in Field]?: string }
      : never
    : never;

/**
 * Turns the type from a list of `Authorization` rules like this:
 *
 * ```typescript
 * [
 *  allow.public(),
 *  allow.owner().inField('otherfield'),
 *  allow.multipleOwners().inField('editors')
 * ]
 * ```
 *
 * Into an object type that includes all auth fields like this:
 *
 * ```typescript
 * {
 *  owner?: string | undefined;
 *  otherfield?: string | undefined;
 *  editors?: string[] | undefined;
 * }
 * ```
 */
export type ImpliedAuthFields<T extends Authorization<any, any, any>> =
  ImpliedAuthField<T> extends never
    ? never
    : UnionToIntersection<ImpliedAuthField<T>>;

export const accessData = <T extends Authorization<any, any, any>>(
  authorization: T,
) => authorization[__data];
