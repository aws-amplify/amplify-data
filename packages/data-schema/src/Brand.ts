const brand = Symbol('brand');
type brand = typeof brand;

export type Brand<T, TBrand extends string> = T & {
  [brand]: TBrand;
};

export function brandTarget<T extends object, TBrand extends string>(
  target: T,
  brandName: TBrand,
): Brand<T, TBrand> {
  return { ...target, [brand]: brandName };
}
