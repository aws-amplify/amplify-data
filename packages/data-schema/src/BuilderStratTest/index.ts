export {
  RefType as RefDataBranding,
  ref as refBranding,
  getData as getDataBranding,
} from './RefDataBranding';
export {
  RefType as RefDataKBranding,
  ref as refKBranding,
  getData as getDataKBranding,
} from './RefDataClassBranding';
export {
  RefType as RefInternalCasting,
  InternalRef as InternalRefInternalCasting,
  ref as refInternalCasting,
} from './RefInternalCasting';
