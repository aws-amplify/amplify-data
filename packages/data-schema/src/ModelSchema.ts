import type {
  DerivedApiDefinition,
  SetTypeSubArg,
} from '@aws-amplify/data-schema-types';
import {
  type ModelType,
  type ModelTypeParamShape,
  type InternalModel,
  isSchemaModelType,
  SchemaModelType,
} from './ModelType';
import type { EnumType, EnumTypeParamShape } from './EnumType';
import type { CustomType, CustomTypeParamShape } from './CustomType';
import type {
  CustomOperation,
  CustomOperationParamShape,
  InternalCustom,
} from './CustomOperation';
export { __auth } from './ModelField';
import { processSchema } from './SchemaProcessor';
import { Authorization } from './Authorization';

/*
 * Notes:
 *
 * TSC output diagnostics to benchmark
 */

type SchemaContent =
  | ModelType<ModelTypeParamShape, any>
  | CustomType<CustomTypeParamShape>
  | EnumType<EnumTypeParamShape>
  | CustomOperation<CustomOperationParamShape, any>;

type ModelSchemaContents = Record<string, SchemaContent>;
type InternalSchemaModels = Record<
  string,
  InternalModel | EnumType<any> | CustomType<any> | InternalCustom
>;

type DatabaseType = 'SQL' | 'DynamoDB';

export type SchemaConfig = {
  databaseType: DatabaseType;
};

export type ModelSchemaParamShape = {
  types: ModelSchemaContents;
  authorization: Authorization<any, any, any>[];
};

type ModelSchemaData<C extends SchemaConfig> = {
  // Differentiate content type based on SQL vs non-SQL
  types: C['databaseType'] extends 'SQL'
    ? ModelSchemaContents
    : ModelSchemaContents;
  authorization: Authorization<any, any, any>[];
};

export type InternalSchema = {
  data: {
    types: InternalSchemaModels;
    authorization: Authorization<any, any, any>[];
  };
};

export type ModelSchema<
  T extends ModelSchemaParamShape,
  UsedMethods extends 'authorization' = never,
> = Omit<
  {
    authorization: <AuthRules extends Authorization<any, any, any>>(
      auth: AuthRules[],
    ) => ModelSchema<
      SetTypeSubArg<T, 'authorization', AuthRules[]>,
      UsedMethods | 'authorization'
    >;
  },
  UsedMethods
> & {
  data: T;
  models: {
    [TypeKey in keyof T['types']]: T['types'][TypeKey] extends ModelType<ModelTypeParamShape>
      ? SchemaModelType<T['types'][TypeKey]>
      : never;
  };
  transform: () => DerivedApiDefinition;
};

/**
 * Amplify API Next Model Schema shape
 */
export type ModelSchemaType = ModelSchema<ModelSchemaParamShape>;

/**
 * Filter the schema types down to only include the ModelTypes as SchemaModelType
 *
 * @param schemaContents The object containing all SchemaContent for this schema
 * @returns Only the schemaContents that are ModelTypes, coerced to the SchemaModelType surface
 */
const filterSchemaModelTypes = (
  schemaContents: ModelSchemaContents,
): Record<string, SchemaModelType> => {
  const modelTypes: Record<string, SchemaModelType> = {};

  if (schemaContents) {
    Object.entries(schemaContents).forEach(([key, content]) => {
      if (isSchemaModelType(content)) {
        modelTypes[key] = content;
      }
    });
  }

  return modelTypes;
};

/**
 * Model Schema type guard
 * @param schema - api-next ModelSchema or string
 * @returns true if the given value is a ModelSchema
 */
export const isModelSchema = (
  schema: string | ModelSchemaType,
): schema is ModelSchemaType => {
  return typeof schema === 'object' && schema.data !== undefined;
};

function _schema<T extends ModelSchemaParamShape>(
  types: T['types'],
  config: SchemaConfig,
) {
  const data: ModelSchemaData<typeof config> = { types, authorization: [] };
  const models = filterSchemaModelTypes(data.types);

  const transform = (): DerivedApiDefinition => {
    const internalSchema: InternalSchema = { data } as InternalSchema;

    return processSchema({ schema: internalSchema });
  };

  const authorization = (rules: any): any => {
    data.authorization = rules;
    return { data, transform, models } as any;
  };

  return {
    data,
    transform,
    authorization,
    models,
  } as ModelSchema<T>;
}

export function bindConfigToSchema(config?: Partial<SchemaConfig>) {
  return <Types extends ModelSchemaContents>(
    types: Types,
  ): ModelSchema<{ types: Types; authorization: [] }> => {
    return _schema(types, {
      ...{ databaseType: 'DynamoDB' },
      ...(config ?? {}),
    });
  };
}

/**
 * The API and data model definition for Amplify Data. Pass in `{ <NAME>: a.model(...) }` to create a database table
 * and exposes CRUDL operations via an API.
 * @param types The API and data model definition
 * @returns An API and data model definition to be deployed with Amplify (Gen 2) experience (`processSchema(...)`)
 * or with the Amplify Data CDK construct (`@aws-amplify/data-construct`)
 */
export const schema = bindConfigToSchema();
