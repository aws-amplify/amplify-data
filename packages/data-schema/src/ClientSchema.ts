import { type __modelMeta__ } from '@aws-amplify/data-schema-types';
import { type GenericModelSchema, type RDSModelSchema } from './ModelSchema';

// MappedTypes
import type { ResolveSchema, SchemaTypes } from './MappedTypes/ResolveSchema';
import type {
  CreateImplicitModelsFromRelations,
  ResolveFieldProperties,
  ResolveStaticFieldProperties,
} from './MappedTypes/ResolveFieldProperties';
import type {
  ModelIdentifier,
  ModelSecondaryIndexes,
  RelationalMetadata,
} from './MappedTypes/ModelMetadata';
import type {
  ExtractNonModelTypes,
  NonModelTypesShape,
} from './MappedTypes/ExtractNonModelTypes';
import {
  ResolveCustomOperations,
  CustomOperationHandlerTypes,
} from './MappedTypes/CustomOperations';

export type ClientSchema<Schema extends GenericModelSchema<any>> =
  InternalClientSchema<Schema>;

/**
 * Types for unwrapping generic type args into client-consumable types
 *
 * @typeParam Schema - Data schema builder model type
 *
 * The following params are used solely as variables in order to simplify mapped type usage.
 * They should not receive external type args.
 *
 * @internal @typeParam NonModelTypes - Custom Types, Enums, and Custom Operations
 * @internal @typeParam ResolvedSchema - Resolve the schema types used by other generics
 * @internal @typeParam ImplicitModels - The implicit models created to represent relationships
 * @internal @typeParam ImplicitModelsIdentifierMeta - The implicite model identifiers derived from ImplicitModels
 * @internal @typeParam ResolvedFields - Resolved client-facing types used for CRUDL response shapes
 * @internal @typeParam SecondaryIndexes - Map of model secondary index metadata
 */
type InternalClientSchema<
  Schema extends GenericModelSchema<any>,
  NonModelTypes extends NonModelTypesShape = ExtractNonModelTypes<Schema>,
  ResolvedSchema = ResolveSchema<Schema>,
  ImplicitModels = Schema extends RDSModelSchema<any, any>
    ? object
    : CreateImplicitModelsFromRelations<ResolvedSchema>,
  ImplicitModelsIdentifierMeta = {
    [ImplicitModel in keyof ImplicitModels]: {
      identifier: 'id';
    };
  },
  ResolvedFields extends Record<
    string,
    unknown
  > = Schema extends RDSModelSchema<any, any>
    ? ResolveStaticFieldProperties<Schema, NonModelTypes, object>
    : ResolveFieldProperties<Schema, NonModelTypes, ImplicitModels>,
> = InternalClientSchemaEntries<Schema, NonModelTypes, ResolvedFields> & {
  [__modelMeta__]: InternalClientSchemaMetadata<
    Schema,
    NonModelTypes,
    ResolvedSchema,
    ImplicitModelsIdentifierMeta,
    ResolvedFields
  >;
};

/**
 * Types for expanding into client-consumable types
 *
 * @typeParam Schema - Data schema builder model type
 * @typeParam NonModelTypes - Custom Types, Enums, and Custom Operations
 * @typeParam ResolvedFields - Resolved client-facing types used for CRUDL response shapes
 */
type InternalClientSchemaEntries<
  Schema extends GenericModelSchema<any>,
  NonModelTypes extends NonModelTypesShape,
  ResolvedFields extends Record<string, unknown>,
> = CustomOperationHandlerTypes<
  ResolveCustomOperations<
    Schema,
    ResolvedFields,
    NonModelTypes
  >['customOperations']
> &
  ResolvedFields;

/**
 * Types for expanding into client-consumable metadata types
 *
 * @typeParam Schema - Data schema builder model type
 *
 * The following params are used solely as variables in order to simplify mapped type usage.
 * They should not receive external type args.
 *
 * @internal @typeParam NonModelTypes - Custom Types, Enums, and Custom Operations
 * @internal @typeParam ResolvedSchema - Resolve the schema types used by other generics
 * @internal @typeParam ImplicitModelsIdentifierMeta - The implicite model identifiers derived from ImplicitModels
 * @internal @typeParam ResolvedFields - Resolved client-facing types used for CRUDL response shapes
 * @internal @typeParam IdentifierMeta - Resolve the identifier fields for all models
 * @internal @typeParam SecondaryIndexes - Map of model secondary index metadata
 */
type InternalClientSchemaMetadata<
  Schema extends GenericModelSchema<any>,
  NonModelTypes extends NonModelTypesShape,
  ResolvedSchema,
  ImplicitModelsIdentifierMeta,
  ResolvedFields extends Record<string, unknown>,
  IdentifierMeta extends Record<string, any> = ModelIdentifier<
    SchemaTypes<Schema>
  >,
  SecondaryIndexes extends Record<string, any> = Schema extends RDSModelSchema<
    any,
    any
  >
    ? object
    : ModelSecondaryIndexes<SchemaTypes<Schema>>,
> = {
  [__modelMeta__]: IdentifierMeta &
    ImplicitModelsIdentifierMeta &
    SecondaryIndexes &
    RelationalMetadata<ResolvedSchema, ResolvedFields, IdentifierMeta> &
    NonModelTypes &
    ResolveCustomOperations<Schema, ResolvedFields, NonModelTypes>;
};
