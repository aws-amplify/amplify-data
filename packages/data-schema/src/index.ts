import { schema } from './ModelSchema';
import { model } from './ModelType';
import {
  id,
  string,
  integer,
  float,
  boolean,
  date,
  time,
  datetime,
  timestamp,
  email,
  json,
  phone,
  url,
  ipAddress,
} from './ModelField';
import { ref } from './RefType';
import { hasOne, hasMany, belongsTo, manyToMany } from './ModelRelationalField';
import { allow } from './Authorization';
import { customType } from './CustomType';
import { enumType } from './EnumType';
import { query, mutation, subscription } from './CustomOperation';
import { handler } from './Handler';

// We are re-exporting with this pattern to make the JSDoc happy.
// Otherwise the JSDoc wouldn't consistently show up as a customer types
// "a."
export {
  schema,
  model,
  ref,
  customType,
  enumType as enum,
  query,
  mutation,
  subscription,
  hasOne,
  hasMany,
  belongsTo,
  manyToMany,
  allow,
  id,
  string,
  integer,
  float,
  boolean,
  date,
  time,
  datetime,
  timestamp,
  email,
  json,
  phone,
  url,
  ipAddress,
  handler,
};
