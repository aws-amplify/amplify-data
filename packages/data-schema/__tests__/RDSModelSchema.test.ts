import { expectTypeTestsToPassAsync } from 'jest-tsd';
import { configure } from '../src/ModelSchema';
import { a } from '../src/index';

// evaluates type defs in corresponding test-d.ts file
it('should not produce static type errors', async () => {
  await expectTypeTestsToPassAsync(__filename);
});

describe('RDSModelSchema', () => {
  describe('.relationships() modifier', () => {
    it('generates expected schema for unidirectional hasMany', () => {
      const schema = configure({
        database: {
          engine: 'mysql',
          connectionUri: 'fake' as any,
        },
      })
        .schema({
          Post: a.model({
            title: a.string().required(),
            content: a.string(),
            parentBlogId: a.string(),
          }),
          Blog: a.model({
            title: a.string().required(),
            description: a.string(),
          }),
        })
        .authorization([a.allow.public()])
        .relationships((models) => [
          models.Blog.addRelationships({
            childPosts: a.hasMany('Post', 'parentBlogId'),
          }),
        ]);

      expect(schema.transform().schema).toMatchSnapshot();
    });

    it('generates expected schema for bidirectional hasMany<->belongsTo', () => {
      const schema = configure({
        database: {
          engine: 'mysql',
          connectionUri: 'fake' as any,
        },
      })
        .schema({
          Post: a.model({
            title: a.string().required(),
            content: a.string(),
            parentBlogId: a.string(),
          }),
          Blog: a.model({
            title: a.string().required(),
            description: a.string(),
          }),
        })
        .authorization([a.allow.public()])
        .relationships((models) => [
          models.Post.addRelationships({
            parentBlog: a.belongsTo('Blog', 'parentBlogId'),
          }),
          models.Blog.addRelationships({
            childPosts: a.hasMany('Post', 'parentBlogId'),
          }),
        ]);

      expect(schema.transform().schema).toMatchSnapshot();
    });

    it('generates expected schema for unidirectional hasOne', () => {
      const schema = configure({
        database: {
          engine: 'mysql',
          connectionUri: 'fake' as any,
        },
      })
        .schema({
          Supplier: a.model({
            name: a.string().required(),
          }),
          Account: a.model({
            email: a.string().required(),
            supplierId: a.string(),
          }),
        })
        .authorization([a.allow.public()])
        .relationships((models) => [
          models.Supplier.addRelationships({
            account: a.hasOne('Account', 'supplierId'),
          }),
        ]);

      expect(schema.transform().schema).toMatchSnapshot();
    });

    it('generates expected schema for unidirectional belongsTo', () => {
      const schema = configure({
        database: {
          engine: 'mysql',
          connectionUri: 'fake' as any,
        },
      })
        .schema({
          Supplier: a.model({
            name: a.string().required(),
          }),
          Account: a.model({
            email: a.string().required(),
            supplierId: a.string(),
          }),
        })
        .authorization([a.allow.public()])
        .relationships((models) => [
          models.Account.addRelationships({
            supplier: a.belongsTo('Supplier', 'supplierId'),
          }),
        ]);

      expect(schema.transform().schema).toMatchSnapshot();
    });

    it('generates expected schema for bidirectional hasOne<->belongsTo', () => {
      const schema = configure({
        database: {
          engine: 'mysql',
          connectionUri: 'fake' as any,
        },
      })
        .schema({
          Supplier: a.model({
            name: a.string().required(),
          }),
          Account: a.model({
            email: a.string().required(),
            supplierId: a.string(),
          }),
        })
        .authorization([a.allow.public()])
        .relationships((models) => [
          models.Account.addRelationships({
            supplier: a.belongsTo('Supplier', 'supplierId'),
          }),
          models.Supplier.addRelationships({
            account: a.hasOne('Account', 'supplierIds'),
          }),
        ]);

      expect(schema.transform().schema).toMatchSnapshot();
    });

    it('is unusable with non-RDS schema', () => {
      expect(() =>
        a
          .schema({
            Post: a.model({
              title: a.string().required(),
              content: a.string(),
            }),
          })
          // @ts-expect-error .relationships() should not be available for non-RDS schema
          .relationships((models) => []),
      ).toThrow();
    });
  });
});
