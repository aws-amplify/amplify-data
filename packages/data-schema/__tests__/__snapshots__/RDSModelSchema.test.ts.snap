// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RDSModelSchema .relationships() modifier generates expected schema for bidirectional hasMany<->belongsTo 1`] = `
"type Blog @model(timestamps: null) @auth(rules: [{allow: public, provider: apiKey}])
{
  title: String!
  description: String
  childPosts: [Post] @hasMany(references: ["parentBlogId"])
}

type Post @model(timestamps: null) @auth(rules: [{allow: public, provider: apiKey}])
{
  title: String!
  content: String
  parentBlogId: String
  parentBlog: Blog @belongsTo(references: ["parentBlogId"])
}"
`;

exports[`RDSModelSchema .relationships() modifier generates expected schema for bidirectional hasOne<->belongsTo 1`] = `
"type Account @model(timestamps: null) @auth(rules: [{allow: public, provider: apiKey}])
{
  email: String!
  supplierId: String
  supplier: Supplier @belongsTo(references: ["supplierId"])
}

type Supplier @model(timestamps: null) @auth(rules: [{allow: public, provider: apiKey}])
{
  name: String!
  account: Account @hasOne(references: ["supplierIds"])
}"
`;

exports[`RDSModelSchema .relationships() modifier generates expected schema for unidirectional belongsTo 1`] = `
"type Account @model(timestamps: null) @auth(rules: [{allow: public, provider: apiKey}])
{
  email: String!
  supplierId: String
  supplier: Supplier @belongsTo(references: ["supplierId"])
}

type Supplier @model(timestamps: null) @auth(rules: [{allow: public, provider: apiKey}])
{
  name: String!
}"
`;

exports[`RDSModelSchema .relationships() modifier generates expected schema for unidirectional hasMany 1`] = `
"type Blog @model(timestamps: null) @auth(rules: [{allow: public, provider: apiKey}])
{
  title: String!
  description: String
  childPosts: [Post] @hasMany(references: ["parentBlogId"])
}

type Post @model(timestamps: null) @auth(rules: [{allow: public, provider: apiKey}])
{
  title: String!
  content: String
  parentBlogId: String
}"
`;

exports[`RDSModelSchema .relationships() modifier generates expected schema for unidirectional hasOne 1`] = `
"type Account @model(timestamps: null) @auth(rules: [{allow: public, provider: apiKey}])
{
  email: String!
  supplierId: String
}

type Supplier @model(timestamps: null) @auth(rules: [{allow: public, provider: apiKey}])
{
  name: String!
  account: Account @hasOne(references: ["supplierId"])
}"
`;
